### PR TITLE
RavenDB-20304 draft

### PR DESCRIPTION
**Pull request by kalczur**
draft

### CODE REVIEW

File: src/Raven.Studio/typescript/common/generalUtils.ts [hunk 1]
```diff
@@ -740,6 +740,14 @@ class genUtils {
         
         return Array.from(result.values());
     }
+
+    static async tryHandleSubmit<T>(promise: () => Promise<T>) {
+        try {
+            return await promise();
+        } catch (_) {
+            // empty by design
```
Comment (ml054) [inline src/Raven.Studio/typescript/common/generalUtils.ts]: let's at least log to console with console.warn or console.error inside catch.

File: src/Raven.Studio/typescript/components/common/shell/setup.ts [hunk 2]
```diff
@@ -44,3 +45,38 @@ export function initRedux() {
         topology.nodes.subscribe(onClusterTopologyChanged);
     });
 }
+
+function initYup() {
```
Comment (ml054) [inline src/Raven.Studio/typescript/components/common/shell/setup.ts:49]: please make sure it is executed in storybook and tests, ok ? 

If not - please move to diferent class and init it there.

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+    const getClientConfigurationCallback = useAsyncCallback(manageServerService.getClientConfiguration);
+    const getClientGlobalConfigurationCallback = useAsync(manageServerService.getGlobalClientConfiguration, []);
+
+    const { isClusterAdminOrClusterNode: canNavigateToServerSettings } = useAccessManager();
+
+    const { handleSubmit, control, formState, setValue, reset, getValues } = useForm<ClientConfigurationFormData>({
+        resolver: clientConfigurationYupResolver,
+        mode: "onChange",
+        defaultValues: async () =>
+            ClientConfigurationUtils.mapToFormData(await getClientConfigurationCallback.execute(db), false),
```
Comment (ml054) [inline src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx:38]: let's talk about that

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    const globalConfig = useMemo(() => {
+        const globalConfigResult = getClientGlobalConfigurationCallback.result;
+        if (!globalConfigResult) {
+            return null;
+        }
+        return ClientConfigurationUtils.mapToFormData(globalConfigResult, true);
+    }, [getClientGlobalConfigurationCallback.result]);
+
+    const {
+        overrideConfig,
```
Comment (ml054) [inline src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx]: let's talk

Comment (ml054) [reply src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx]: use form

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+import ClientConfigurationUtils from "components/common/clientConfiguration/ClientConfigurationUtils";
+
+interface ClientDatabaseConfigurationProps {
+    db: database;
+}
+
+// TODO: show modal on exit intent if is dirty
+export default function ClientDatabaseConfiguration({ db }: ClientDatabaseConfigurationProps) {
+    const { manageServerService } = useServices();
+    const getClientConfigurationCallback = useAsyncCallback(manageServerService.getClientConfiguration);
```
Comment (ml054) [inline src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx:29]: can we remove that?

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+                    </Card>
+                )}
+
+                <Card className="card flex-row p-2 mt-4">
+                    <Row className="flex-grow-1">
+                        <Col>
+                            <FormCheckbox
+                                control={control}
+                                name="identityPartsSeparatorEnabled"
+                                afterChange={(event) =>
```
Comment (ml054) [inline src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx]: let's use watch .

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+                                        placeholder="Enter seed number"
+                                        disabled={!loadBalancerSeedEnabled || !overrideConfig}
+                                    />
+                                </Col>
+                            </Row>
+                        </Col>
+                        {globalConfig && (
+                            <>
+                                <Col>
+                                    <Input value={globalConfig.loadBalancerSeedValue || undefined} disabled />
```
Comment (ml054) [inline src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx]: `??`

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+                                Read balance behavior
+                            </FormCheckbox>
+                        </Col>
+                        <Col>
+                            <FormSelect
+                                control={control}
+                                name="readBalanceBehaviorValue"
+                                disabled={!readBalanceBehaviorEnabled || !overrideConfig}
+                            >
+                                <FormSelectOption<ReadBalanceBehavior> label="None" value="None" />
```
Comment (ml054) [inline src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx]: what about empty option? (-select-)

Comment (ml054) [inline src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx]: use map from list.

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx (hunk 3 above)
Comment (kalczur) [reply src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx:38]: formState has no information whether defaultValues ​​was successfully set. If it fails default values is set to empty object

option 1: use useAsyncCallback()
option 2: write a custom hook wrapping useForm() in an additional field, e.g. isLoadError

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx (hunk 8 above)
Comment (kalczur) [reply src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx]: It's reactstrap so I guess you'd need to add an option
`<option label="(-select-)" value={null} />`

or I can add e.g. `showEmptyOption` props to `<FormSelect />`

File: src/Raven.Studio/typescript/components/common/shell/setup.ts (hunk 2 above)
Comment (kalczur) [reply src/Raven.Studio/typescript/components/common/shell/setup.ts:49]: I checked and it works everywhere

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx (hunk 5 above)
Comment (kalczur) [reply src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx:29]: https://github.com/ravendb/ravendb/pull/16356#discussion_r1177798760

File: src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        if (!globalConfigResult) {
+            return null;
+        }
+        return ClientConfigurationUtils.mapToFormData(globalConfigResult, true);
+    }, [getClientGlobalConfigurationCallback.result]);
+
+    const onSave: SubmitHandler<ClientConfigurationFormData> = async (formData) => {
+        tryHandleSubmit(async () => {
+            await manageServerService.saveClientConfiguration(ClientConfigurationUtils.mapToDto(formData, false), db);
+            reset(null, { keepValues: true });
```
Comment (ml054) [inline src/Raven.Studio/typescript/components/pages/database/settings/clientConfiguration/ClientDatabaseConfiguration.tsx:54]: do we want to put this reset into tryhandlesubmit?

**Review by ml054 — APPROVED** [review]

**Pull request closed without merging**